src/main/java/seedu/binbash/storage/Storage.java

```diff
@@ -108,78 +177,101 @@ private ArrayList<String> readTxtFile() throws BinBashException, IOException {
+            try {
+                Item item = convertLineToItem(line);
+                if (item != null) {
+                    outputList.add(item);
+                } else {
                     isCorrupted = true;
                 }
-                parseState = READING_IN_PROFIT_QUANTITIES;
-                break;
-            case "setprofit":
-                assert parseState == READING_IN_PROFIT_QUANTITIES;
-                handleSetProfitString(commandArgs, itemManager);
-                parseState = READING_IN_ITEM;
-                break;
-            default:
+            } catch (NumberFormatException | DateTimeParseException e) {
                 isCorrupted = true;
             }
         }
+        return outputList;
     }
 
     /**
-     * Handles the "add" command by parsing the arguments and executing the command to add an item to the item list.
-     *
-     * @param addCommandParser The parser object used to parse the add command.
-     * @param commandArgs The arguments of the add command.
-     * @param itemManager The ItemList object to which the new item is added.
-     * @throws InvalidCommandException if the command is invalid.
-     * @throws ParseException if there is an error in parsing the command arguments.
-     */
-    private void handleAddCommand(AddCommandParser addCommandParser, String[] commandArgs, ItemList itemManager)
-            throws InvalidCommandException, ParseException {
-        Command addCommand = addCommandParser.parse(commandArgs);
-        addCommand.execute(itemManager);
-    }
-
-    /**
-     * Handles the "setprofit" string by updating the profit-related properties of the most recently added item.
+     * Parses a single line from the text file to create an item object.
      *
-     * @param commandArgs The arguments of the setprofit command.
-     * @param itemManager The ItemList object containing the recently added item.
+     * @param line The string representation of a single line in the text file.
+     * @return The item object created from the line, or {@code null} if the item type is unknown.
      */
-    private void handleSetProfitString(String[] commandArgs, ItemList itemManager) {
-        List<Item> itemList = itemManager.getItemList();
-        Item recentlyAddedItem = itemList.get(itemList.size() - 1);
-
-        String totalUnitsPurchased = commandArgs[TOTAL_UNITS_PURCHASED_INDEX];
-        recentlyAddedItem.setTotalUnitsPurchased(Integer.parseInt(totalUnitsPurchased));
-
-        if (recentlyAddedItem instanceof RetailItem) {
-            String totalUnitsSold = commandArgs[TOTAL_UNITS_SOLD_INDEX];
-            ((RetailItem) recentlyAddedItem).setTotalUnitsSold(Integer.parseInt(totalUnitsSold));
+    protected Item convertLineToItem(String line) {
+        String[] itemElements = line.split("\\|");
+        String itemType = itemElements[ITEM_TYPE_INDEX];
+
+        switch (itemType) {
+        case "OperationalItem":
+
```

Comment:
The long arguments that are passed into the constructor here are mostly repeated for all the items, thus making your switch case look long. Perhaps before the flow enters your switch case, you can define these long arguments as shorter variables, and pass these variables into your constructor so that it can reduce the length of your switch case and make it look more readable. Nothing major, just to improve code readability.

```suggestion
String itemName = itemElements[ITEM_NAME_INDEX];
String itemDescription = itemElements[ITEM_DESCRIPTION_INDEX];
int itemQuantity = Integer.parseInt(itemElements[ITEM_QUANTITY_INDEX]);
LocalDate itemExpirationDate = LocalDate.parse(itemElements[ITEM_EXPIRATION_DATE_INDEX], EXPECTED_INPUT_DATE_FORMAT);
double itemCostPrice = Double.parseDouble(itemElements[ITEM_COST_PRICE_INDEX]);
double itemSalePrice = Double.parseDouble(itemElements[ITEM_SALE_PRICE_INDEX]);
int itemThreshold = Integer.parseInt(itemElements[ITEM_THRESHOLD_INDEX]);
```